Change indexer to use blobstore abstraction

Bento007 opened the pull request:
Change indexer to use blobstore abstractions and updated associated s3 test cases. Added stub for gcp file indexer.

ttung commented on dss/events/handlers/index.py:
```diff
@@ -17,32 +15,52 @@
-        else:
-            logger.debug(f"Not indexing S3 creation event for key: {key}")
+        bucket_name = event['Records'][0]["s3"]["bucket"]["name"]
+        blobstore = S3BlobStore()
```
I would encourage you to call `Config.get_cloud_specific_handles(..)`

Bento007 replied:
I just recently found that code. Wish i knew about it sooner. I'll make this change. 👍

ttung commented on dss/events/handlers/index.py:
```diff
@@ -17,32 +15,52 @@
+        # TODO retreive bucket_name and key from gs_event
+        bucket_name = None
+        key = None
+        blobstore = GSBlobStore(os.environ["GOOGLE_APPLICATION_CREDENTIALS"])
```
ditto

ttung commented on dss/events/handlers/index.py:
```diff
@@ -17,32 +15,52 @@
+        raise
+
+
+def process_new_indexable_object(blobstore, bucket_name, key, replica, logger) -> None:
```
you could actually type most of these parameters.

ttung replied:
Also, if you're passing in replica here, you can just get the BlobStore object inside `process_new_indexable_object`

Bento007 commented:
I also added blobstore to test_indexer and infra. I intended to make this it's own PR, but I forgot to create new branch and added the changes to this branch. Sorry about that.

ttung commented on dss/events/handlers/index.py:
```diff
@@ -6,43 +6,58 @@
         raise
 
 
+def process_new_gs_indexable_object(event, logger) -> None:
```
Did I miss the place where this is called?

Bento007 replied:
I'll remove for now and add it in my next PR, since test coverage is not there. Otherwise it won't pass the code coverage.

ttung commented on dss/events/handlers/index.py:
```diff
@@ -52,19 +67,18 @@ def is_bundle_to_index(key) -> bool:
 
-def read_bundle_manifest(s3, bucket_name, bundle_key, logger):
-    manifest_string = s3.Object(bucket_name, bundle_key).get()['Body'].read().decode("utf-8")
+def read_bundle_manifest(blobstore, bucket_name, bundle_key, logger):
```
https://github.com/HumanCellAtlas/data-store/issues/390

Bento007 replied:
This will be a separate PR

ttung commented on tests/infra.py:
```diff
@@ -132,39 +132,29 @@ def __getattr__(self, item: str) -> typing.Any:
+        self.name = os.path.basename(object_summary)
+        self.path = object_summary
         self.uuid = str(uuid.uuid4())
+        self.url = f"s3://{bundle.bucket}/{self.path}"
```
If `TestFile` is not to be cloud-specific, then this is not the right source url.

Bento007 replied:
You're absolutely right.

ttung commented on tests/test_indexer.py:
```diff
@@ -74,7 +73,12 @@ class TestIndexer(unittest.TestCase, DSSAsserts, StorageTestSupport):
 
     @classmethod
     def setUpClass(cls):
+        cls.replica = "aws"
```
is this intended? if you're making this cloud-agnostic, shouldn't this test both code paths?

Bento007 replied:
That is intended for now. I'm working on making a base class for test_indexer to test any cloud service we use. That will be in my next PR.

ttung replied:
Cool, sgtm.

Bento007 commented:
Ready for review!

ttung commented on tests/infra.py:
```diff
@@ -132,39 +133,30 @@ def __getattr__(self, item: str) -> typing.Any:
+        self.name = os.path.basename(object_summary)
+        self.path = object_summary
         self.uuid = str(uuid.uuid4())
+        self.url = f"{Config.get_storage_schema(replica)}://{bundle.bucket}/{self.path}"
```
I believe that as a group, we've decided not to allow complicated expressions (anything more than a variable or a variable+index) inside a f-string.

Bento007 replied:
I'll switch to using string concatenation instead of f-string.
